### PR TITLE
Introduce Prometheus Java Client integration test

### PR DESCRIPTION
**Pull request by rickie**
There are some XXX's left to resolve, but apart from those, I think the PR is ready to be reviewed!

For those XXX's I'll still work on creating a PR for them.

### CODE REVIEW

**Comment by rickie** [comment]
It's successful!

File: integration-tests/prometheus-java-client-init.patch [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                         </path>
+                         <!-- Other annotation processors go here.
+ 
+--- a/prometheus-metrics-exporter-opentelemetry/src/test/java/io/prometheus/metrics/exporter/opentelemetry/PrometheusInstrumentationScopeTest.java
++++ b/prometheus-metrics-exporter-opentelemetry/src/test/java/io/prometheus/metrics/exporter/opentelemetry/PrometheusInstrumentationScopeTest.java
+@@ -7,6 +7,8 @@ import org.junit.jupiter.api.Test;
+ 
+ class PrometheusInstrumentationScopeTest {
+ 
++  // XXX: Introduce support for rewriting `assertThatThrownBy#havingRootCause()`.
```
Comment (rickie) [inline integration-tests/prometheus-java-client-init.patch:61]: Filed: https://github.com/PicnicSupermarket/error-prone-support/pull/1471

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit. One point of attention: there appears to be a flaky test:
```
[ERROR] io.prometheus.metrics.core.metrics.SlidingWindowTest.rotate -- Time elapsed: 0.023 s <<< FAILURE!
org.opentest4j.AssertionFailedError:
[Start time: 1734782202868, current time: 1734782232878, elapsed time: 30010]
expected: [1.0, 1.0, 1.0, 1.0, 1.0]
 but was: [1.0, 1.0, 1.0, 1.0]
    at io.prometheus.metrics.core.metrics.SlidingWindowTest$Observer.assertValues(SlidingWindowTest.java:30)
    at io.prometheus.metrics.core.metrics.SlidingWindowTest.rotate(SlidingWindowTest.java:57)
    at java.base/java.lang.reflect.Method.invoke(Method.java:569)
    at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
    at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

@mohamedsamehsalah if you review #1471 then we can strike one precondition off the list :)

File: integration-tests/prometheus-java-client-init.patch [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
++++ b/prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/Quantiles.java
+@@ -41,6 +41,8 @@ public class Quantiles implements Iterable<Quantile> {
+    * Create a new Quantiles instance. You can either create Quantiles with one of the static {@code
+    * Quantiles.of(...)} methods, or you can use the {@link Quantiles#builder()}.
+    */
++  // XXX: Don't suggest changes when there are local methods with the same name.
++  @SuppressWarnings("ExplicitArgumentEnumeration")
+   public static Quantiles of(Quantile... quantiles) {
+     return of(Arrays.asList(quantiles));
+   }
```
[the remark this answers is not in the document]
Comment (rickie) [reply integration-tests/prometheus-java-client-init.patch:58]: If not, I'll try to make time for it on Monday! Today won't work for me.

**Comment by rickie** [comment]
Found a few flaky test issues on their GitHub about the same class as what you noticed: https://github.com/prometheus/client_java/issues/956, https://github.com/prometheus/client_java/pull/946. 

I also experienced the issue on my local machine. Maybe for now we should report and exclude the specific case via `-Dtest`?

File: integration-tests/prometheus-java-client-init.patch (hunk 2 above)
[the remark this answers is not in the document]
Comment (Stephan202) [reply integration-tests/prometheus-java-client-init.patch]: I filed #1475 and rebased this PR to exclude these suppressions from the patch file.

**Comment by Stephan202** [comment]
> I also experienced the issue on my local machine. Maybe for now we should report and exclude the specific case via `-Dtest`?

Yeah, agree. I filed prometheus/client_java#1242 and will reference that issue in an accompanying comment.

**Comment by Stephan202** [comment]
Rebased and added a commit. The PR now requires #1475 and #1476 to be merged first. Open points (not today):
- Cleanup of `.github/workflows/run-integration-tests.yml`.
- A closer look at the generated changes.
- A closer look at the `@SuppressWarnings("AssertThatThrownBy")` in the initial patch file.

File: .github/workflows/run-integration-tests.yml [hunk 3]
```diff
@@ -34,12 +36,13 @@ jobs:
             objects.githubusercontent.com:443
             oss.sonatype.org:443
             raw.githubusercontent.com:443
+            registry-1.docker.io:443
```
Comment (Stephan202) [inline .github/workflows/run-integration-tests.yml:39]: Added based on [this report](https://app.stepsecurity.io/github/PicnicSupermarket/error-prone-support/actions/runs/12464380205?jobid=34788363557&tab=network-events).

**Comment by Stephan202** [comment]
I added a commit that assumes #1477.

**Comment by Stephan202** [comment]
Added a commit to document some possible follow-up work. Open points:
- Fixing the build. Somehow pulling Docker images keeps failing:
  ```
  [main] INFO tc.testcontainers/ryuk:0.11.0 - Pulling docker image: testcontainers/ryuk:0.11.0. Please be patient; this may take some time but only needs to be done once.
  [docker-java-stream--1295187651] ERROR com.github.dockerjava.api.async.ResultCallbackTemplate - Error during callback
  com.github.dockerjava.api.exception.InternalServerErrorException: Status 500: {"message":"Head \"[https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/0.11.0\](https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/0.11.0/)": Get \"https://auth.docker.io/token?account=githubactions&scope=repository%3Atestcontainers%2Fryuk%3Apull&service=registry.docker.io\": dial tcp 54.185.253.63:443: connect: connection refused"}
  ```
- Cleanup of `.github/workflows/run-integration-tests.yml`

**Comment by rickie** [comment]
Nice work @Stephan202 🥳 !

**Review by Stephan202 — APPROVED** [review]
With the latest commits, stuff should work :)

Suggested commit message:
```
Introduce Prometheus Java Client integration test (#1468)

And document some possible future improvements.
```

**Comment by rickie** [comment]
Restarted the builds because of transient build failures related to Jitpack. Will merge once 🟢  😄

**Pull request merged**